pkg/controller/test-images-distributor/test_images_distributor.go

```diff
@@ -0,0 +1,364 @@
+	return &handler.EnqueueRequestsFromMapFunc{ToRequests: handler.ToRequestsFunc(
+		func(mo handler.MapObject) []reconcile.Request {
+			imageStream, ok := mo.Object.(*imagev1.ImageStream)
+			if !ok {
+				logrus.WithField("type", fmt.Sprintf("%T", mo.Object)).Error("Got object that was not an ImageStram")
+				return nil
+			}
+			var requests []reconcile.Request
+			for _, imageStreamTag := range imageStream.Spec.Tags {
+				// Not sure why this happens but seems to be a thing
+				if imageStreamTag.Name == "" {
+					continue
+				}
+				name := types.NamespacedName{
+					Namespace: mo.Meta.GetNamespace(),
+					Name:      fmt.Sprintf("%s:%s", mo.Meta.GetName(), imageStreamTag.Name),
+				}
+				if !filter(name) {
+					continue
+				}
+				// We have to squeeze both the target cluster name and the imageStreamTag name into a reconcile.Request
+				// Internally, this gets put onto the workqueue as a single string in namespace/name notation and split
+				// later on. This means that we can not use a slash as delimiter for the cluster and the namespace.
+				for _, buildCluster := range buildClusters.List() {
+					name := types.NamespacedName{
+						Namespace: buildCluster + clusterAndNamespaceDelimiter + name.Namespace,
+						Name:      name.Name,
+					}
+					requests = append(requests, reconcile.Request{NamespacedName: name})
+				}
+			}
+			return requests
+		},
+	),
+	}
+}
+
+const clusterAndNamespaceDelimiter = "_"
+
+func decodeRequest(req reconcile.Request) (string, types.NamespacedName, error) {
+	clusterAndNamespace := strings.Split(req.Namespace, "_")
+	if n := len(clusterAndNamespace); n != 2 {
+		return "", types.NamespacedName{}, fmt.Errorf("didn't get two but %d segments when trying to extract cluster and namespace", n)
+	}
+	return clusterAndNamespace[0], types.NamespacedName{Namespace: clusterAndNamespace[1], Name: req.Name}, nil
+}
+
+type reconciler struct {
+	ctx                      context.Context
+	log                      *logrus.Entry
+	registryClient           ctrlruntimeclient.Client
+	buildClusterClients      map[string]ctrlruntimeclient.Client
+	pullSecretGetter         func() []byte
+	successfulImportsCounter *prometheus.CounterVec
+	failedImportsCounter     *prometheus.CounterVec
+	dryRun                   bool
+}
+
+func (r *reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithField("request", req.String())
+	log.Info("Starting reconciliation")
+	err := r.reconcile(req, log)
+	if err != nil {
+		log.WithError(err).Error("Reconciliation failed")
+	} else {
+		log.Info("Finished reconciliation")
+	}
+	return reconcile.Result{}, controllerutil.SwallowIfTerminal(err)
+}
+
+func (r *reconciler) reconcile(req reconcile.Request, log *logrus.Entry) error {
+	cluster, decoded, err := decodeRequest(req)
+	if err != nil {
+		return fmt.Errorf("failed to decode request %s: %w", req, err)
+	}
+	// Propagate the cluster field back up
+	*log = *log.WithField("cluster", cluster)
```

Comment:
What's with the pointer dereferences here?

Reply:
It sets the value that the pointer we get points to, thus propagating the new field up the logger in `Reconcile` so when we log errors there we have the cluster as a field